crates/ruma-common/src/events/room/member.rs

```diff
@@ -377,45 +395,126 @@ fn membership_change(
 }
 
 impl OriginalRoomMemberEvent {
+    /// Obtain the details about this event that are required to calculate a membership change.
+    ///
+    /// This is required when you want to calculate the change a redacted `m.room.member` event
+    /// made.
+    pub fn details(&self) -> MembershipDetails<'_> {
+        self.content.details()
+    }
+
+    fn prev_details(&self) -> Option<MembershipDetails<'_>> {
+        self.unsigned.prev_content.as_ref().map(|c| c.details())
+    }
+
     /// Helper function for membership change.
     ///
     /// Check [the specification][spec] for details.
     ///
     /// [spec]: https://spec.matrix.org/v1.2/client-server-api/#mroommember
     pub fn membership_change(&self) -> MembershipChange {
-        membership_change(
-            &self.content,
-            self.unsigned.prev_content.as_ref(),
-            &self.sender,
-            &self.state_key,
-        )
+        membership_change(self.details(), self.prev_details(), &self.sender, &self.state_key)
     }
 }
 
-impl OriginalSyncStateEvent<RoomMemberEventContent> {
+impl RedactedRoomMemberEvent {
+    /// Obtain the details about this event that are required to calculate a membership change.
+    ///
+    /// This is required when you want to calculate the change a redacted `m.room.member` event
+    /// made.
+    pub fn details(&self) -> MembershipDetails<'_> {
+        self.content.details()
+    }
+
     /// Helper function for membership change.
     ///
+    /// Since redacted events don't have `unsigned.prev_content`, you have to pass the `.details()`
+    /// of the previous `m.room.member` event manually (if three is a previous `m.room.member`
+    /// event).
+    ///
     /// Check [the specification][spec] for details.
     ///
     /// [spec]: https://spec.matrix.org/v1.2/client-server-api/#mroommember
-    pub fn membership_change(&self) -> MembershipChange {
-        membership_change(
-            &self.content,
-            self.unsigned.prev_content.as_ref(),
-            &self.sender,
-            &self.state_key,
-        )
+    pub fn membership_change(
+        &self,
+        prev_details: Option<MembershipDetails<'_>>,
+    ) -> MembershipChange {
+        membership_change(self.details(), prev_details, &self.sender, &self.state_key)
     }
 }
 
-impl StrippedStateEvent<RoomMemberEventContent> {
+impl OriginalSyncRoomMemberEvent {
+    /// Obtain the details about this event that are required to calculate a membership change.
+    ///
+    /// This is required when you want to calculate the change a redacted `m.room.member` event
+    /// made.
+    pub fn details(&self) -> MembershipDetails<'_> {
+        self.content.details()
+    }
+
+    fn prev_details(&self) -> Option<MembershipDetails<'_>> {
```

Comment:
Same as the previous one, any reason why not make this pub?

Reply:
I think having a public `prev_content` method (`self.unsigned.prev_content.as_ref()`) could be useful, but not this. It's artificially limited to a small number of fields of the actual content.